port-javascript: Makefile overhaul, sdk, and user mods

pmp-p opened the pull request:
because emcc looks by itself into that ports folders cache but clang -E cannot.

This way if a header is missing clang used as preprocessor won't tap by default in /usr/include.

It will break build and will force user to explicitly add 
-I$(EM_CACHE)/asmjs/ports-builds/****somelib*****/include for his libs.

Commit:
port-javascript: prevent clang as a parser to use host headers

because emcc looks by itself into that ports folders cache but clang -E cannot.

This way if a header is missing clang used as preprocessor won't tap by default in /usr/include.

It will break build and will force user to explicitly add 
-I$(EM_CACHE)/asmjs/ports-builds/****somelib*****/include for his libs.

dpgeorge commented:
As an alternative to this, I tried defining `CPP = emcc -E` but this doesn't like multiple input files, so I then concat'd all files in to one and ran that through `emcc -E`, and it seemed to work.  I don't know if that would be preferred (using `emcc -E` instead of `clang -E`) but it would need a modification to `py/mkrules.mk` to do the concat'ing.

pmp-p commented:
emcc sure does ** a lot **  of things clang alone cannot, including  ... downloading ports source code, patching and building them on the fly for the selected targets. and i maybe wrong but think prepro is run **before** emcc giving no chances to do all that.

also there's all vital -s FLAGS clang  will not undesrtand.

+1 for emcc everywhere it could be used, or building modules will be hotline hell added to loss of control for debugging / ram settings / library switches.

embeddedt commented:
> +1 for emcc everywhere it could be used, or building modules will be hotline hell added to loss of control for debugging / ram settings / library switches.

By no means am I a MicroPython expert, but when I was fiddling with this to get LittlevGL working with JavaScript, I found that the `clang -E` hack was only necessary at the beginning of the build process, for QSTR generation. `emcc` was used everywhere else.

That means that debugging settings and extra libraries shouldn't be affected.

pmp-p commented:
> That means that debugging settings and extra libraries shouldn't be affected.
indeed debugging flags seem harmles, but extra libraries have an influence with modules building system because eg -s USE_SDL=2 or -s SOCKET_WEBRTC=0/1 would get passed via the CFLAGS_EXTRA and breaks clang.

in that case i think clang cannot be used as an emcc replacement at all. it would complicate things too much for user module creation. What if emscripten changes some vars or the undocumented cache path or use a relative path in EM_CACHE ?

embeddedt commented:
>  What if emscripten changes some vars or the undocumented cache path or use a relative path in EM_CACHE ?

In that case, `EM_CACHE` could quickly be updated.

pmp-p commented:
@embeddedt  i see trouble handling a bunch of modules vs system emscripten / emsdk git that would use different path ....

dpgeorge commented:
I tried to use `emcc -E` instead of `clang -E` but it just cannot work, because:
- `emcc -E` must only be passed .c files, not .h files (and also cannot read from stdin)
- concatenating all required inputs to `emcc -E` into another temporary file and then passing that to `emcc -E` doesn't work because it can't find header includes (which are sometimes relative to the source file being processed, and if everything is combined into a single .c file then it's impossible to know where the relative headers are)

It seems the only way to fix it properly would be to patch `emcc` (which is really out of scope here).

So I guess the fix in this PR is the only way forward.

pmp-p commented:
then maybe fix this + find a normalized way to add  @embeddedt 's method for settings includes for modules, maybe  CPPFLAGS_USERMOD along with - but instead  - of CFLAGS_USERMOD which should keep emcc specifics.

Maybe a prefilled list of CPPFLAGS_USERMOD with expected standard emscripten paths : there's not so much ports after all.

embeddedt commented:
@pmp-p 

>  downloading ports source code, patching and building them on the fly for the selected targets. and i maybe wrong but think prepro is run before emcc giving no chances to do all that.

Correct, but have you heard of `embuilder.py`? It can be run separately from `emcc` and could be used to fetch ports before `clang` is invoked manually.

pmp-p commented:
@embeddedt thanks that would seem very usefull to easy preload things for user modules
```
$ embuilder.py build --lto --pic sdl2
embuilder:INFO: building and verifying sdl2
embuilder:INFO: ...success
```

Commit:
readability + ensure prepro. knows about target

dpgeorge commented:
@pmp-p I see you made some more additions.  Is this good to go in?

pmp-p commented:
@dpgeorge i succeed to build micropython+lvgl based on those options, so i think it's quite ok for a start.

 Still there's the problem to get something readable for user mods flags but that can come later, as @embeddedt  is working actively on that too. i think later settings CPPFLAGS and calling embuild from an emsripten.mk ( ie could be cheerp.mk there's not only emscripten around ) , as it is very compiler specific will be next step.

embeddedt commented:
@dpgeorge 

>  Is this good to go in?

At the moment, sadly, I see little choice but to merge this, even though it's quite non-portable (i.e. assumes that the cache is in `~/.emscripten_cache` and that ports are built).

I'm thinking that someone should raise an issue on the Emscripten end at some point to point out this flaw with `emcc`. Ideally we shouldn't need to invoke `clang` directly at all.

@pmp-p 

> @embeddedt is working actively on that too

I am?

pmp-p commented:
@embeddedt , sorry maybe i mistook some comments but i had vaguely the idea that you would have integrated LittleVGL as a module via the user mod build system if it wasn't for the root pointer problem.

embeddedt commented:
@pmp-p You would have to discuss that with @amirgon - I wasn't the one who implemented LittlevGL's integration in the first place. I'm just making it work with JavaScript/Emscripten.

embeddedt commented on ports/javascript/Makefile:
```diff
@@ -16,7 +16,10 @@ INC += -I$(BUILD)
+	CPP += -D__EMSCRIPTEN__
+	CPP += --sysroot $(EMSCRIPTEN)/system
+	CPP += -isystem $(EMSCRIPTEN)/system/include/libc
+	CPP += -cxx-isystem $(EMSCRIPTEN)/system/include/libcxx
```
@dpgeorge @pmp-p

I've finally discovered that there is a slightly better way of handling these :

```
CPP += $(addprefix -isystem, $(shell $(CC) -E -x c++ /dev/null -v 2>&1 |sed -e '/^\#include <...>/,/^End of search/{ //!b };d'))
```

By parsing the verbose output of `emcc` it's possible to retrieve the include paths.

pmp-p replied:
Very nice ! i like that it's a very complete output,

@embeddedt  one thing though : it seems SDL include folder looks like a ```-s USE_SDL=1``` (1.3.0) instead of a 2

also maybe add LC_ALL=C just in case ```End of search``` gets a localized version ?

embeddedt replied:
To fix that the `JSFLAGS` line should be moved above `CC = emcc -g4` and the new line is this:

```
$(addprefix -isystem, $(shell env LC_ALL=C $(CC) -E -x c++ /dev/null -v 2>&1 |sed -e '/^\#include <...>/,/^End of search/{ //!b };d'))
```

embeddedt replied:
Confirmed that the hack is not needed.

pmp-p replied:
it works very well when adding specific -s USE_flags that are normally in CFLAGS but should not be passed to clang -E. 

 i see a reason for user mod to have separate CFLAGS and CPPFLAGS
or some more magic in the makefile to remove the  -s * pattern from the user mod flags ?

embeddedt replied:
What is the problem? It is compiling fine without any errors about unknown options.

pmp-p replied:
@embeddedt you're right, i missed that user flags include will get translated to -isystem ones too

i'll try to enhance that PR for user mod system !

embeddedt replied:
Ideally we also want to pass proper predefined macros to `clang` as well; I'm working on another one of these parsing hacks that should do that. That way we can avoid this ugliness (which will fall apart on a different architecture):

https://github.com/littlevgl/lv_micropython/blob/1381be71fb2f11c6e0841b2beebf1f3aa84bcba2/ports/javascript/Makefile#L26

embeddedt replied:
Alright, I can now pass macros to `clang`; see my latest comments on the `Makefile` (near `CPP = clang -E`).

embeddedt commented on ports/javascript/Makefile:
```diff
@@ -16,7 +16,10 @@ INC += -I$(BUILD)
 CPP = clang -E
```
To pass macros manually this needs to be:

```suggestion
CPP = clang -E -undef
```

embeddedt replied:
For some reason, GitHub refuses to let me comment on the remaining lines of the file.

```patch
-SRC_QSTR += $(SRC_C)
+SRC_QSTR += $(BUILD)/clang_predefs.h $(SRC_C)
+
+$(BUILD)/clang_predefs.h:
+       @emcc $(JSFLAGS) -E -x c /dev/null -dM > $@
```

Commit:
get includes via idea of @embeddedt

imho good start, as it can be upgraded later to support user mods requirements.
And is far better than stopping build on GNU/Linux systems.

Though i'm not sure that solution will work on some emsdk platforms if they lack decent shell.

Somebody would need to test mac/win

embeddedt commented on ports/javascript/Makefile:
```diff
@@ -13,13 +13,29 @@ INC += -I.
+	CPP += --sysroot $(EMSCRIPTEN)/system	
+	CPP += $(addprefix -isystem, $(shell env LC_ALL=C $(CC) $(CFLAGS_EXTRA) -E -x c++ /dev/null -v 2>&1 |sed -e '/^\#include <...>/,/^End of search/{ //!b };d'))
+	# Act like 'emcc'
+	CPP += -U__i386 -U__i386 -Ui386 -U__SSE -U__SSE_MATH -U__SSE2 -U__SSE2_MATH -U__MMX__ -U__SSE__ -U__SSE_MATH__ -U__SSE2__ -U__SSE2_MATH__
```
The manual undefining of macros should be replaced by the trick I described here: https://github.com/micropython/micropython/pull/4875#pullrequestreview-259447603

pmp-p replied:
@embeddedt maybe push the all the tricks at once first here https://github.com/littlevgl/lv_micropython/blob/pmpp_test/ports/javascript/Makefile so i can try them with child modules ?

embeddedt replied:
They were already there: https://github.com/littlevgl/lv_micropython/commit/3bd1eabc7fd3551ce8fe35dfc12b492d5ce1f998

pmp-p replied:
iirc i tried those but i had a problem with macro redef, but adding -Wno-macro-redefined to $(CPP) seems to fix it

embeddedt replied:
Disabling that warning is not a good solution to the problem (it just hides it). What macros are you getting the warning on?

pmp-p replied:
```
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
Including User C Module from cmod/example
Makefile:178: warning: overriding recipe for target 'clean'
../../py/mkrules.mk:157: warning: ignoring old recipe for target 'clean'
Makefile:300: warning: overriding recipe for target 'micropython.html'
../../py/mkrules.mk:128: warning: ignoring old recipe for target 'micropython.html'
mkdir -p build/genhdr
GEN build/genhdr/mpversion.h
GEN build/genhdr/moduledefs.h
GEN build/genhdr/qstr.i.last
build/clang_predefs.h:24:9: error: '__CLANG_ATOMIC_LLONG_LOCK_FREE' macro redefined [-Werror,-Wmacro-redefined]
#define __CLANG_ATOMIC_LLONG_LOCK_FREE 1
        ^
<built-in>:300:9: note: previous definition is here
#define __CLANG_ATOMIC_LLONG_LOCK_FREE 2
        ^
build/clang_predefs.h:85:9: error: '__GCC_ATOMIC_LLONG_LOCK_FREE' macro redefined [-Werror,-Wmacro-redefined]
#define __GCC_ATOMIC_LLONG_LOCK_FREE 1
        ^
<built-in>:310:9: note: previous definition is here
#define __GCC_ATOMIC_LLONG_LOCK_FREE 2
        ^
build/clang_predefs.h:107:9: error: '__INT64_C_SUFFIX__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT64_C_SUFFIX__ LL
        ^
<built-in>:175:9: note: previous definition is here
#define __INT64_C_SUFFIX__ L
        ^
build/clang_predefs.h:108:9: error: '__INT64_FMTd__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT64_FMTd__ "lld"
        ^
<built-in>:173:9: note: previous definition is here
#define __INT64_FMTd__ "ld"
        ^
build/clang_predefs.h:109:9: error: '__INT64_FMTi__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT64_FMTi__ "lli"
        ^
<built-in>:174:9: note: previous definition is here
#define __INT64_FMTi__ "li"
        ^
build/clang_predefs.h:110:9: error: '__INT64_MAX__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT64_MAX__ 9223372036854775807LL
        ^
<built-in>:207:9: note: previous definition is here
#define __INT64_MAX__ 9223372036854775807L
        ^
build/clang_predefs.h:111:9: error: '__INT64_TYPE__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT64_TYPE__ long long int
        ^
<built-in>:172:9: note: previous definition is here
#define __INT64_TYPE__ long int
        ^
build/clang_predefs.h:117:9: error: '__INTMAX_C_SUFFIX__' macro redefined [-Werror,-Wmacro-redefined]
#define __INTMAX_C_SUFFIX__ LL
        ^
<built-in>:64:9: note: previous definition is here
#define __INTMAX_C_SUFFIX__ L
        ^
build/clang_predefs.h:118:9: error: '__INTMAX_FMTd__' macro redefined [-Werror,-Wmacro-redefined]
#define __INTMAX_FMTd__ "lld"
        ^
<built-in>:62:9: note: previous definition is here
#define __INTMAX_FMTd__ "ld"
        ^
build/clang_predefs.h:119:9: error: '__INTMAX_FMTi__' macro redefined [-Werror,-Wmacro-redefined]
#define __INTMAX_FMTi__ "lli"
        ^
<built-in>:63:9: note: previous definition is here
#define __INTMAX_FMTi__ "li"
        ^
build/clang_predefs.h:120:9: error: '__INTMAX_MAX__' macro redefined [-Werror,-Wmacro-redefined]
#define __INTMAX_MAX__ 9223372036854775807LL
        ^
<built-in>:42:9: note: previous definition is here
#define __INTMAX_MAX__ 9223372036854775807L
        ^
build/clang_predefs.h:121:9: error: '__INTMAX_TYPE__' macro redefined [-Werror,-Wmacro-redefined]
#define __INTMAX_TYPE__ long long int
        ^
<built-in>:61:9: note: previous definition is here
#define __INTMAX_TYPE__ long int
        ^
build/clang_predefs.h:125:9: error: '__INTPTR_MAX__' macro redefined [-Werror,-Wmacro-redefined]
#define __INTPTR_MAX__ 2147483647L
        ^
<built-in>:46:9: note: previous definition is here
#define __INTPTR_MAX__ 9223372036854775807L
        ^
build/clang_predefs.h:127:9: error: '__INTPTR_WIDTH__' macro redefined [-Werror,-Wmacro-redefined]
#define __INTPTR_WIDTH__ 32
        ^
<built-in>:79:9: note: previous definition is here
#define __INTPTR_WIDTH__ 64
        ^
build/clang_predefs.h:136:9: error: '__INT_FAST64_FMTd__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT_FAST64_FMTd__ "lld"
        ^
<built-in>:280:9: note: previous definition is here
#define __INT_FAST64_FMTd__ "ld"
        ^
build/clang_predefs.h:137:9: error: '__INT_FAST64_FMTi__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT_FAST64_FMTi__ "lli"
        ^
<built-in>:281:9: note: previous definition is here
#define __INT_FAST64_FMTi__ "li"
        ^
build/clang_predefs.h:138:9: error: '__INT_FAST64_MAX__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT_FAST64_MAX__ 9223372036854775807LL
        ^
<built-in>:279:9: note: previous definition is here
#define __INT_FAST64_MAX__ 9223372036854775807L
        ^
build/clang_predefs.h:139:9: error: '__INT_FAST64_TYPE__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT_FAST64_TYPE__ long long int
        ^
<built-in>:278:9: note: previous definition is here
#define __INT_FAST64_TYPE__ long int
        ^
build/clang_predefs.h:152:9: error: '__INT_LEAST64_FMTd__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT_LEAST64_FMTd__ "lld"
        ^
<built-in>:240:9: note: previous definition is here
#define __INT_LEAST64_FMTd__ "ld"
        ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
../../py/mkrules.mk:74: recipe for target 'build/genhdr/qstr.i.last' failed
make: *** [build/genhdr/qstr.i.last] Error 1
make: *** Deleting file 'build/genhdr/qstr.i.last'
```

embeddedt replied:
CPP has to be `clang -E -undef`.

pmp-p replied:
indeed, seems better

pmp-p replied:
hmm on my port i get that with -undef , while -Wno-macro-redefined pass the build
```CC ../../extmod/moductypes.c
../../extmod/moductypes.c:747:19: error: use of undeclared identifier 'MP_QSTR_SHORT'
    { MP_ROM_QSTR(MP_QSTR_SHORT), MP_ROM_INT(TYPE2SMALLINT(INT16, 4)) },
                  ^
../../extmod/moductypes.c:748:19: error: use of undeclared identifier 'MP_QSTR_USHORT'
    { MP_ROM_QSTR(MP_QSTR_USHORT), MP_ROM_INT(TYPE2SMALLINT(UINT16, 4)) },
                  ^
../../extmod/moductypes.c:751:19: error: use of undeclared identifier 'MP_QSTR_INT'; did you mean 'MP_QSTR_INT8'?
    { MP_ROM_QSTR(MP_QSTR_INT), MP_ROM_INT(TYPE2SMALLINT(INT32, 4)) },
                  ^~~~~~~~~~~
                  MP_QSTR_INT8
../../py/obj.h:248:40: note: expanded from macro 'MP_ROM_QSTR'
#define MP_ROM_QSTR(q) MP_OBJ_NEW_QSTR(q)
                                       ^
../../py/obj.h:92:56: note: expanded from macro 'MP_OBJ_NEW_QSTR'
#define MP_OBJ_NEW_QSTR(qst) ((mp_obj_t)((((mp_uint_t)(qst)) << 2) | 2))
                                                       ^
build/genhdr/qstrdefs.generated.h:297:6: note: 'MP_QSTR_INT8' declared here
QDEF(MP_QSTR_INT8, (const byte*)"\xce\x04" "INT8")
     ^
../../extmod/moductypes.c:752:19: error: use of undeclared identifier 'MP_QSTR_UINT'; did you mean 'MP_QSTR_UINT8'?
    { MP_ROM_QSTR(MP_QSTR_UINT), MP_ROM_INT(TYPE2SMALLINT(UINT32, 4)) },
                  ^~~~~~~~~~~~
                  MP_QSTR_UINT8
../../py/obj.h:248:40: note: expanded from macro 'MP_ROM_QSTR'
#define MP_ROM_QSTR(q) MP_OBJ_NEW_QSTR(q)
                                       ^
../../py/obj.h:92:56: note: expanded from macro 'MP_OBJ_NEW_QSTR'
#define MP_OBJ_NEW_QSTR(qst) ((mp_obj_t)((((mp_uint_t)(qst)) << 2) | 2))
                                                       ^
build/genhdr/qstrdefs.generated.h:309:6: note: 'MP_QSTR_UINT8' declared here
QDEF(MP_QSTR_UINT8, (const byte*)"\xbb\x05" "UINT8")
     ^
../../extmod/moductypes.c:755:19: error: use of undeclared identifier 'MP_QSTR_LONG'; did you mean 'MP_QSTR_VOID'?
    { MP_ROM_QSTR(MP_QSTR_LONG), MP_ROM_INT(TYPE2SMALLINT(INT32, 4)) },
                  ^~~~~~~~~~~~
                  MP_QSTR_VOID
../../py/obj.h:248:40: note: expanded from macro 'MP_ROM_QSTR'
#define MP_ROM_QSTR(q) MP_OBJ_NEW_QSTR(q)
                                       ^
../../py/obj.h:92:56: note: expanded from macro 'MP_OBJ_NEW_QSTR'
#define MP_OBJ_NEW_QSTR(qst) ((mp_obj_t)((((mp_uint_t)(qst)) << 2) | 2))
                                                       ^
build/genhdr/qstrdefs.generated.h:310:6: note: 'MP_QSTR_VOID' declared here
QDEF(MP_QSTR_VOID, (const byte*)"\x31\x04" "VOID")
     ^
../../extmod/moductypes.c:756:19: error: use of undeclared identifier 'MP_QSTR_ULONG'
    { MP_ROM_QSTR(MP_QSTR_ULONG), MP_ROM_INT(TYPE2SMALLINT(UINT32, 4)) },
                  ^
../../extmod/moductypes.c:762:19: error: use of undeclared identifier 'MP_QSTR_LONGLONG'
    { MP_ROM_QSTR(MP_QSTR_LONGLONG), MP_ROM_INT(TYPE2SMALLINT(INT64, 4)) },
                  ^
../../extmod/moductypes.c:763:19: error: use of undeclared identifier 'MP_QSTR_ULONGLONG'; did you mean
      'MP_QSTR_ENOTCONN'?
    { MP_ROM_QSTR(MP_QSTR_ULONGLONG), MP_ROM_INT(TYPE2SMALLINT(UINT64, 4)) },
                  ^~~~~~~~~~~~~~~~~
                  MP_QSTR_ENOTCONN
../../py/obj.h:248:40: note: expanded from macro 'MP_ROM_QSTR'
#define MP_ROM_QSTR(q) MP_OBJ_NEW_QSTR(q)
                                       ^
../../py/obj.h:92:56: note: expanded from macro 'MP_OBJ_NEW_QSTR'
#define MP_OBJ_NEW_QSTR(qst) ((mp_obj_t)((((mp_uint_t)(qst)) << 2) | 2))
                                                       ^
build/genhdr/qstrdefs.generated.h:287:6: note: 'MP_QSTR_ENOTCONN' declared here
QDEF(MP_QSTR_ENOTCONN, (const byte*)"\x79\x08" "ENOTCONN")
     ^
../../extmod/moductypes.c:771:8: error: invalid application of 'sizeof' to an incomplete type
      'const mp_rom_map_elem_t []'
STATIC MP_DEFINE_CONST_DICT(mp_module_uctypes_globals, mp_module_uctypes_globals_table);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../py/obj.h:314:21: note: expanded from macro 'MP_DEFINE_CONST_DICT'
            .used = MP_ARRAY_SIZE(table_name), \
                    ^~~~~~~~~~~~~~~~~~~~~~~~~
../../py/misc.h:111:33: note: expanded from macro 'MP_ARRAY_SIZE'
#define MP_ARRAY_SIZE(a) (sizeof(a) / sizeof((a)[0]))
                                ^~~
../../extmod/moductypes.c:771:8: error: invalid application of 'sizeof' to an incomplete type
      'const mp_rom_map_elem_t []'
STATIC MP_DEFINE_CONST_DICT(mp_module_uctypes_globals, mp_module_uctypes_globals_table);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../py/obj.h:315:22: note: expanded from macro 'MP_DEFINE_CONST_DICT'
            .alloc = MP_ARRAY_SIZE(table_name), \
                     ^~~~~~~~~~~~~~~~~~~~~~~~~
../../py/misc.h:111:33: note: expanded from macro 'MP_ARRAY_SIZE'
#define MP_ARRAY_SIZE(a) (sizeof(a) / sizeof((a)[0]))
                                ^~~
10 errors generated.
shared:ERROR: compiler frontend failed to generate LLVM bitcode, halting
../../py/mkrules.mk:47: recipe for target 'build/extmod/moductypes.o' failed
make: *** [build/extmod/moductypes.o] Error 1
```

embeddedt replied:
`-Wno-macro-redefined` is just hiding the issue because we don't want any predefined macros coming from `clang`.

Can you clean and rebuild?

pmp-p replied:
my clang was missing some macro cleared by -undef, but ```-include $(BUILD)/clang_predefs.h``` can re add them.

Commit:
Update Makefile

Commit:
flags readability

Commit:
inject emcc macro into clang

embeddedt commented:
@pmp-p I have two questions:

* Why is the `else` clause necessary on the `ifdef EMSCRIPTEN` clause? I don't see a case where someone would want to compile the JavaScript port *without* Emscripten. If anything I would think that that should produce an error with something like `$(error Cannot find Emscripten SDK)` rather than choosing the host `clang` or `gcc`.
* Is the `ASYNC` part of the `Makefile` relevant to this PR?

pmp-p commented:
Primary goal of this PR is to have a decent building process for people with user modules. and prevent *emscripten* from doing stupid things not in one but in all its use case :

1 ) emscripten can be used or not.

- There's not only emscripten available around for compiling to asm.js or wasm.
- Compiling to native/cross the javascript port can help testing code validity for other ports and that the emscripten specific blocks are correcly isolated via ifdef so it should be possible to use plain gcc or clang to build.

2)  if emscripent is used, it can be used async or not.
- async is relevant - even if crippled on purpose here - because it can help some people stuck with some issues already open around. emterpreter can't be Ignored but I certainly would not open a PR for that because i think it's not a solution, i don't even want to advertise it because it prevents people from searching solutions and upstream them (or at least try to).

Of course anyone can feel free to remove any feature or put comments in front, or reject the whole PR because it's GNU/Linux or mac  only,
But i don't think javascript port has any future if it stays crippled or uncared for ( once heat wave is passed you won't see me for a while ).

embeddedt committed:
Ensure `clang_predefs.h` is included in QSTR extraction

embeddedt committed:
Update Makefile

Commit:
Merge pull request #1 from embeddedt/patch-1

Ensure `clang_predefs.h` is included in QSTR extraction

Commit:
no more needed with the new makefile trick

embeddedt commented:
@dpgeorge I think this one is ready to be merged. We've been using it for quite a while on `lv_micropython` and I haven't seen any new issues.

dpgeorge commented:
>  I think this one is ready to be merged.

Ok, but it has changed a lot (added a lot) since the first review and seems now to do a few unrelated things.  So I'll need to rereview it in detail.

embeddedt commented:
> Ok, but it has changed a lot (added a lot) since the first review and seems now to do a few unrelated things. So I'll need to rereview it in detail.

That's fine.

The new stuff essentially extracts the Emscripten predefined macros and include paths automatically and passes them to `clang` instead of the previous method where we had to use hardcoded include paths.

pmp-p commented:
meanwhile a problem arose with the new upstreaming emscripten branch
```
emsdk install 1.38.40-upstream
emsdk activate 1.38.40-upstream
```
will lead to 
```
CC main.c
CC mphalport.c
CC modutime.c
LINK build/firmware.js
shared:ERROR: BINARYEN_TRAP_MODE is not supported by the LLVM wasm backend
Makefile:59: recipe for target 'build/micropython.js' failed
make: *** [build/micropython.js] Error 1
```
also a new option made it in emscripten for coroutines with a lot less performance loss than emterpreting see https://kripken.github.io/blog/wasm/2019/07/16/asyncify.html

embeddedt commented:
> meanwhile a problem arose with the new upstreaming emscripten branch

Cursory research suggests that this shouldn't be (too) difficult to fix. I'll have a look.

> in addition the LDFLAG `-Wl,-Map=$@.map,--cref -Wl,--gc-sections` is unsupported by clang 10 wasm-ld linker.

This is just for generating binary maps and doesn't affect the compiled output at all, so it can safely be removed or replaced with the correct flags.

Commit:
handle emscripten-upstream branch and add a default source path for firefox debugger

embeddedt commented on ports/javascript/Makefile:
```diff
@@ -22,6 +26,12 @@ ifdef EMSCRIPTEN
+	#check if not using emscripten-upstream branch
+	ifeq (,$(findstring upstream/bin, $(EMMAKEN_COMPILER)))
+		JSFLAGS += -s "BINARYEN_TRAP_MODE='clamp'"
+		LDFLAGS = -m32 -Wl,-Map=$@.map,--cref -Wl,--gc-sections
```
Why not `+=` on `LDFLAGS` in order to pick up the `-m32` from line 18?

Commit:
Update Makefile

pmp-p commented:
@embeddedt , removing seems ok, please have a look

Commit:
Update Makefile

Commit:
missing tab

Commit:
Update Makefile

Commit:
use DEBUG flag like other ports

Commit:
Update Makefile

embeddedt committed:
Fix issues with Makefile

embeddedt committed:
Add CFLAGS_MOD

embeddedt committed:
Make unused function warning produce error

Commit:
Merge pull request #2 from littlevgl/pmpp_6_fix

Fix issues with Makefile

Commit:
start a layout handling clang/wasi

Commit:
handle makefile choice for running micropython

Commit:
Update Makefile

Commit:
fix regression for frozen

Commit:
add shared so emscripten people have a look at setjmp error

```emmake make SHARED=1```
should lead to
```Fatal: GOT.func entry with no import/export: $emscripten_longjmp_jmpbuf```

pmp-p commented:
it seems emcc is maybe expected to behave strictly as "clang -E" as a preprocessor from what @sbc100 said there https://github.com/emscripten-core/emscripten/issues/9317#issuecomment-528102182 even though i was wrong on symptoms ( that is files args no parameters args) .

So maybe just file an issue on https://github.com/emscripten-core/emscripten/issues instead of cooking magic in the makefile.

embeddedt commented:
@pmp-p I don't have time at the moment to further investigate why `emcc` wasn't behaving, but if you want to open an issue, I'll help where I can.

pmp-p commented:
@embeddedt  ho thx , maybe pack a small repro case of the behaviour @dpgeorge observed

pmp-p commented:
issue is now open https://github.com/emscripten-core/emscripten/issues/9466

thanks @andrewleech.

embeddedt commented:
Thank you @andrewleech! Hopefully this issue gets resolved soon.

pmp-p commented:
#5140 affects that PR so either it is not good either there's a problem in QSTR parser used this way.

pmp-p commented:
issue https://github.com/emscripten-core/emscripten/issues/9353 can also drastically help to simplify that PR in the (near?) future and unification with other compilers.

still #5084 should allow a hook slot for user to alter build behaviour regarding his sdk tools himself. frozen codes and user modules are NOT part of MicroPython framework : they are user's pipeline.
eg transpiled code like https://github.com/pmp-p/micropython-ports-wasm/blob/master/cmod/embed.pym or f-string converter to apply last moment on code shared with cpython or pyodide.

User mod build system may not be able to cover cases over than basic ones ( which is still _good_ for beginners but may be questionnable from other categories).

jedie commented:
Any news here?

dpgeorge commented:
The issue with Emscripten preprocessing was fixed by 08e0e065f4fa26cb1f52567ad53052310bc656e6 and c13853f4dab9c65ee6d1af99b25f999b16ef2827.  The latter commit makes this PR obsolete.
